### ISSUE TITLE
NullPointerException in PushWoosh.<init>

### ISSUE THREAD

**Issue by n-belokopytov** [Pushwoosh/pushwoosh-android-sdk#74]
PW 5.1.1, FCM 11.4.2 - No further info available, since it pretty much breaks the initialisation of the app.

```

Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'com.pushwoosh.notification.e com.pushwoosh.g.d()' on a null object reference
--
1 | com.pushwoosh.Pushwoosh.<init> | Unknown Source
2 | com.pushwoosh.Pushwoosh.<clinit> | Unknown Source
... 13 more


```

**Comment by wfhm** [comment]
Hi, checking it.

**Comment by wfhm** [comment]
It looks like you use Pushwoosh SDK methods before initializing Pushwosh library. It can only happen when you call `Pushwoosh.SOME_THING` inside the constructor of your ContentProvider, or in your Application constructor. Could you please check it, and if it is indeed so, use Pushwoosh methods only after `Application.onCreate()`? Handling of this exception will also be added in the next release.

```
public class TestingApp extends Application {
    @Override
    public void onCreate() {
        super.onCreate();
        //Here you can start use Pushwoosh library
    }
}
```

**Comment by n-belokopytov** [comment]
@wfhm we do exactly that. Here is how we do it.

`    private void initPushWoosh() {
        Object service = getSystemService(Context.ACTIVITY_SERVICE);
        if (service != null) {
            List<ActivityManager.RunningAppProcessInfo> runningAppProcesses = ((ActivityManager) service).getRunningAppProcesses();
            if (runningAppProcesses != null && runningAppProcesses.size() != 0) {
                for (ActivityManager.RunningAppProcessInfo runningAppProcessInfo : runningAppProcesses) {
                    boolean isCurrentProcess = runningAppProcessInfo.pid == android.os.Process.myPid();
                    boolean isMainProcessName = getPackageName().equals(runningAppProcessInfo.processName);
                    if (isCurrentProcess && isMainProcessName) {
                        pushWooshFacade.register();
                        break;
                    }
                }
            }
        }
    }
`
```
    @Override
    public void onCreate() {
        super.onCreate();
....
       initPushWoosh();
   }
```

What is up with that btw? I think I've seen it in your docs. So IMHO if you go full "Pushwoosh.getInstance()", it's ultimately up to you to ensure that every time I am trying to use the instance - it's setup and ready or at least doesn't crash. You're misusing the singleton pattern here if you're storing some obscure state in it. You have to know that in many apps there are 10 or 15 calls to initialize different parts of the App in "Application::onCreate" and you need to make sure that the order of the initialisation doesn't matter or else there will be too many side effects - there are literally thousands of third-party libraries that might interfere.

**Comment by wfhm** [comment]
@n-belokopytov 

We use the same approach Firebase uses for its library initialization:

https://firebase.googleblog.com/2016/12/how-does-firebase-initialize-on-android.html

Content providers are initialized before Application is created (before onCreate) - that guarantees Pushwoosh lib initialization in application context.

**Comment by wfhm** [comment]
@n-belokopytov 

By any chance could you provide us with a full stack trace?

**Comment by n-belokopytov** [comment]
@wfhm sorry, I pasted everything I had.

**Comment by n-belokopytov** [comment]
Got more stacktraces for you on 5.4.1.

> java.lang.IllegalStateException: Pushwoosh library not initialized. Please use Pushwoosh library after Application.onCreate() method

```
1 | com.getyourguide.tracking.PushWooshFacade.registerOnPushwoosh | PushWooshFacade.java, line 82
-- | -- | --
2 | com.getyourguide.tracking.PushWooshFacade.ensureRegistration | PushWooshFacade.java, line 65
3 | com.getyourguide.android.GYGApplication.initPushWoosh | GYGApplication.java, line 175
4 | com.getyourguide.android.GYGApplication.onCreate | GYGApplication.java, line 135
5 | android.app.Instrumentation.callApplicationOnCreate | Instrumentation.java, line 1013

6 | android.app.ActivityThread.handleBindApplication | ActivityThread.java, line 4609
7 | android.app.ActivityThread.access$1600 | ActivityThread.java, line 169
8 | android.app.ActivityThread$H.handleMessage | ActivityThread.java, line 1337
9 | android.os.Handler.dispatchMessage | Handler.java, line 102
10 | android.os.Looper.loop | Looper.java, line 136
11 | android.app.ActivityThread.main | ActivityThread.java, line 5476
12 | java.lang.reflect.Method.invokeNative | Native Method
13 | java.lang.reflect.Method.invoke | Method.java, line 515
14 | com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run | ZygoteInit.java, line 1283
15 | com.android.internal.os.ZygoteInit.main | ZygoteInit.java, line 1099
16 | dalvik.system.NativeStart.main | Native Method
Caused by: java.lang.IllegalStateException: Pushwoosh library not initialized. Please use Pushwoosh library after Application.onCreate() method
1 | com.pushwoosh.Pushwoosh.<init> | Unknown Source
2 | com.pushwoosh.Pushwoosh.<clinit>

```

**Comment by wfhm** [comment]
@n-belokopytov Great, thank you! We are investigating it.

**Comment by wfhm** [comment]
@n-belokopytov fixed in [5.5.4](https://github.com/Pushwoosh/pushwoosh-android-sdk/tree/5.5.4)
